### PR TITLE
Report error for native integer token as using name or alias target

### LINKED ISSUE #42975: Report error for native integer token as using alias target
Consider reporting an error for a native integer token as a using alias target:
```C#
using A = nint; // error?

class Program
{
    A F() => default;
}
```

### PR DESCRIPTION
**Pull request by cston**
Fixes #42975

### CODE REVIEW

**Review by 333fred — APPROVED** [review]

File: src/Compilers/CSharp/Portable/Binder/Binder_Symbols.cs [hunk 1]
```diff
@@ -934,12 +933,23 @@ private NamedTypeSymbol BindNativeIntegerSymbolIfAny(IdentifierNameSyntax node,
...
-            if (IsInsideNameof && node.Parent is ArgumentSyntax &&
-                node.Parent?.Parent?.Parent is InvocationExpressionSyntax invocation &&
-                (invocation.Expression as IdentifierNameSyntax)?.Identifier.ContextualKind() == SyntaxKind.NameOfKeyword)
+            if (node.Parent is { } parent)
             {
-                // Don't bind nameof(nint) or nameof(nuint) so that ERR_NameNotInContext is reported.
-                return null;
+                switch (parent.Kind())
+                {
+                    case SyntaxKind.Attribute:
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Binder/Binder_Symbols.cs:940]: >SyntaxKind.Attribute [](start = 25, length = 20)

Consider adding a comment for each new case

Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Binder/Binder_Symbols.cs:940]: >SyntaxKind.Attribute [](start = 25, length = 20)

Consider strengthening this check against any possible future changes to the shape of the AttributeSyntax by checking that the ```node``` is at the ```Name``` position.

File: src/Compilers/CSharp/Portable/Binder/Binder_Symbols.cs [hunk 2]
```diff
@@ -934,12 +933,23 @@ private NamedTypeSymbol BindNativeIntegerSymbolIfAny(IdentifierNameSyntax node,
...
+                    case SyntaxKind.UsingDirective:
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Binder/Binder_Symbols.cs:941]: >case SyntaxKind.UsingDirective: [](start = 20, length = 31)

Please strengthen this check by checking that the node is at the ```Name``` position.

File: src/Compilers/CSharp/Test/Semantic/Semantics/NativeIntegerTests.cs [hunk 3]
```diff
@@ -3507,20 +3507,117 @@ static void verify(CSharpCompilation comp)
...
+using A2 = nuint;
 class Program
 {
-    N F() => default;
+    A1 F1() => default;
+    A2 F2() => default;
 }";
+            var expectedDiagnostics = new[]
+            {
+                // (1,12): error CS0246: The type or namespace name 'nint' could not be found (are you missing a using directive or an assembly reference?)
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Semantic/Semantics/NativeIntegerTests.cs:3519]: >// (1,12): error CS0246: The type or namespace name 'nint' could not be found (are you missing a using directive or an assembly reference?) [](start = 16, length = 139)

Please add a test for a success case, i.e. when the type or namespace can be found.

File: src/Compilers/CSharp/Test/Semantic/Semantics/NativeIntegerTests.cs [hunk 4]
```diff
@@ -3507,20 +3507,117 @@ static void verify(CSharpCompilation comp)
...
+using nuint;
+class Program
+{
+}";
+            var expectedDiagnostics = new[]
+            {
+                // (1,1): hidden CS8019: Unnecessary using directive.
+                // using nint;
+                Diagnostic(ErrorCode.HDN_UnusedUsingDirective, "using nint;").WithLocation(1, 1),
+                // (1,7): error CS0246: The type or namespace name 'nint' could not be found (are you missing a using directive or an assembly reference?)
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Semantic/Semantics/NativeIntegerTests.cs:3549]: >// (1,7): error CS0246: The type or namespace name 'nint' could not be found (are you missing a using directive or an assembly reference?) [](start = 16, length = 138)

Please add a test for a success case, i.e. when the type or namespace can be found.

File: src/Compilers/CSharp/Test/Semantic/Semantics/NativeIntegerTests.cs [hunk 5]
```diff
@@ -3507,20 +3507,117 @@ static void verify(CSharpCompilation comp)
...
+        {
+            var source =
+@"[nint]
+[nuint]
+class Program
+{
+}";
+            var expectedDiagnostics = new[]
+            {
+                // (1,2): error CS0246: The type or namespace name 'nintAttribute' could not be found (are you missing a using directive or an assembly reference?)
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Semantic/Semantics/NativeIntegerTests.cs:3579]: >// (1,2): error CS0246: The type or namespace name 'nintAttribute' could not be found (are you missing a using directive or an assembly reference?) [](start = 16, length = 147)

Please add a test for a success case, i.e. when the type can be found.

**Comment by AlekseyTs** [comment]
> @"using nint = System.Int16;

I assume using ```nint``` as an alias name should be allowed. Please add a test for success case with usage of the alias, probably with escaping identifier at the reference point.

---
Refers to: src/Compilers/CSharp/Test/Semantic/Semantics/NativeIntegerTests.cs:3475 in 9f895ba. [](commit_id = 9f895ba5bd4cb63828fad9653cd643142351312e, deletion_comment = False)

**Comment by AlekseyTs** [comment]
>                 // (4,22): error CS8400: Feature 'native-sized integers' is not available in C# 8.0. Please use language version 9.0 or greater.

Is this a breaking change for a legacy code? Intentional?

---
Refers to: src/Compilers/CSharp/Test/Semantic/Semantics/NativeIntegerTests.cs:3483 in 9f895ba. [](commit_id = 9f895ba5bd4cb63828fad9653cd643142351312e, deletion_comment = False)

**Comment by AlekseyTs** [comment]
Done with review pass (iteration 1)

**Comment by cston** [comment]
>                 // (4,22): error CS8400: Feature 'native-sized integers' is not available in C# 8.0. Please use language version 9.0 or greater.

The error reported is for `nuint` which is not an alias.

---
In reply to: [708088642](https://github.com/dotnet/roslyn/pull/48538#issuecomment-708088642) [](ancestors = 708088642)

---
Refers to: src/Compilers/CSharp/Test/Semantic/Semantics/NativeIntegerTests.cs:3483 in 9f895ba. [](commit_id = 9f895ba5bd4cb63828fad9653cd643142351312e, deletion_comment = False)

**Comment by cston** [comment]
> @"using nint = System.Int16;

This test tests both a success case (`nint`) and failure case (`nuint`).

---
In reply to: [708088363](https://github.com/dotnet/roslyn/pull/48538#issuecomment-708088363) [](ancestors = 708088363)

---
Refers to: src/Compilers/CSharp/Test/Semantic/Semantics/NativeIntegerTests.cs:3475 in 9f895ba. [](commit_id = 9f895ba5bd4cb63828fad9653cd643142351312e, deletion_comment = False)

**Comment by AlekseyTs** [comment]
> @"using nint = System.Int16;

> This test tests both a success case (nint) and failure case (nuint).

Success cases should be compiled, executed, expected runtime behavior should be observed.



---
In reply to: [708639774](https://github.com/dotnet/roslyn/pull/48538#issuecomment-708639774) [](ancestors = 708639774,708088363)

---
Refers to: src/Compilers/CSharp/Test/Semantic/Semantics/NativeIntegerTests.cs:3475 in 9f895ba. [](commit_id = 9f895ba5bd4cb63828fad9653cd643142351312e, deletion_comment = False)

**Comment by AlekseyTs** [comment]
>                 // (4,22): error CS8400: Feature 'native-sized integers' is not available in C# 8.0. Please use language version 9.0 or greater.

> The error reported is for nuint which is not an alias.

Consider separating error and success tests to avoid confusing the reader and simplify maintenance.

---
In reply to: [708635107](https://github.com/dotnet/roslyn/pull/48538#issuecomment-708635107) [](ancestors = 708635107,708088642)

---
Refers to: src/Compilers/CSharp/Test/Semantic/Semantics/NativeIntegerTests.cs:3483 in 9f895ba. [](commit_id = 9f895ba5bd4cb63828fad9653cd643142351312e, deletion_comment = False)

**Comment by cston** [comment]
@AlekseyTs, all feedback has been addressed, thanks.

**Review by AlekseyTs — APPROVED** [review]
LGTM (iteration 3)

**Pull request merged**